pkg/kubelet/images/image_manager.go

```diff
@@ -164,14 +233,33 @@ func (m *imageManager) EnsureImageExists(ctx context.Context, pod *v1.Pod, conta
 		m.backOff.Next(backOffKey, m.backOff.Clock.Now())
 
 		msg, err := evalCRIPullErr(container, imagePullResult.err)
-		return "", msg, err
+		return imageRef, msg, err
 	}
 	m.podPullingTimeRecorder.RecordImageFinishedPulling(pod.UID)
 	imagePullDuration := time.Since(startTime).Truncate(time.Millisecond)
 	m.logIt(ref, v1.EventTypeNormal, events.PulledImage, logPrefix, fmt.Sprintf("Successfully pulled image %q in %v (%v including waiting). Image size: %v bytes.",
 		container.Image, imagePullResult.pullDuration.Truncate(time.Millisecond), imagePullDuration, imagePullResult.imageSize), klog.Info)
 	metrics.ImagePullDuration.WithLabelValues(metrics.GetImageSizeBucket(imagePullResult.imageSize)).Observe(imagePullDuration.Seconds())
 	m.backOff.GC()
+
+	m.lock.Lock()
+	if imagePullResult.pullCredentialsHash == "" {
+		// successful pull no auth hash returned, auth was not required so we should reset the hashmap for this
+		// imageref since auth is no longer required for the local image cache, allowing use of the ImageRef
+		// by other pods if it remains cached and pull policy is PullIfNotPresent
+		delete(m.ensureSecretPulledImage, imageRef)
+	} else {
+		// store/create hashMatch map entry for auth config hash key used to pull the image
+		// for this imageref (digest)
+		digest := m.ensureSecretPulledImage[imageRef]
+		if digest == nil {
+			digest = &ensureSecretPulledImageDigest{HashMatch: make(map[string]*ensuredInfo)}
+			m.ensureSecretPulledImage[imageRef] = digest
```

Comment:
Shouldn't the key of this map be a list of auth digests? A single image could have multiple auths for different users. cc: @mikebrow

Reply:
`ensureSecretPulledImage` is a map `map[string]*ensureSecretPulledImageDigest`.
- The map key is image ref.
- The key is `ensureSecretPulledImageDigest` which contains hashMap `HashMatch map[string]*ensuredInfo`.
  - the key is auth's hash
  - the value is `&ensuredInfo{ensured=true, lastEnsuredDate=time.Now()} `

So a single image can have multi auths here.

Reply:
yeah probably will be easier to read with the proposed name changes.. :-)